platform/src/main/python/dlpx/virtualization/platform/_upgrade.py

```diff
@@ -28,64 +27,83 @@
+                                                      source_config_impl)
             return source_config_impl
+
         return source_config_decorator
 
-    def linked_source(self, migration_id):
+    def linked_source(self,
+                      migration_id,
+                      migration_type=MigrationType.PLATFORM):
         def linked_source_decorator(linked_source_impl):
-            std_mig_id = self.__migration_id_set.add(
-                migration_id, linked_source_impl.__name__)
-            self.linked_source_id_to_impl[std_mig_id] = v.check_function(
-                linked_source_impl, Op.UPGRADE_LINKED_SOURCE)
+            if migration_type == MigrationType.PLATFORM:
+                self.platform_migrations.add_linked_source(
+                    migration_id, linked_source_impl)
+            else:
+                # here migration type has to be lua.
+                self.lua_migrations.add_linked_source(migration_id,
+                                                      linked_source_impl)
             return linked_source_impl
+
         return linked_source_decorator
 
-    def virtual_source(self, migration_id):
+    def virtual_source(self,
+                       migration_id,
+                       migration_type=MigrationType.PLATFORM):
         def virtual_source_decorator(virtual_source_impl):
-            std_mig_id = self.__migration_id_set.add(
-                migration_id, virtual_source_impl.__name__)
-            self.virtual_source_id_to_impl[std_mig_id] = v.check_function(
-                virtual_source_impl, Op.UPGRADE_VIRTUAL_SOURCE)
+            if migration_type == MigrationType.PLATFORM:
+                self.platform_migrations.add_virtual_source(
+                    migration_id, virtual_source_impl)
+            else:
+                # here migration type has to be lua.
+                self.lua_migrations.add_virtual_source(migration_id,
+                                                       virtual_source_impl)
             return virtual_source_impl
+
         return virtual_source_decorator
 
-    def snapshot(self, migration_id):
+    def snapshot(self, migration_id, migration_type=MigrationType.PLATFORM):
         def snapshot_decorator(snapshot_impl):
-            std_mig_id = self.__migration_id_set.add(
-                migration_id, snapshot_impl.__name__)
-            self.snapshot_id_to_impl[std_mig_id] = v.check_function(
-                snapshot_impl, Op.UPGRADE_SNAPSHOT)
+            if migration_type == MigrationType.PLATFORM:
+                self.platform_migrations.add_snapshot(migration_id,
+                                                      snapshot_impl)
+            else:
+                # here migration type has to be lua.
+                self.lua_migrations.add_snapshot(migration_id, snapshot_impl)
             return snapshot_impl
+
         return snapshot_decorator
 
     @property
     def migration_id_list(self):
```

Comment:
Should this be a list of LUA and Python migrations? Would we ever need to get sorted migrations for LUA?

Reply:
you're talking about the migration id list right? actually this was something I wanted to ask others about. basically the migration id list here is used during the build to get the list of all the migration ids. The list is needed because we want to verify when doing the upgrade on the delphix engine that we have the migrations needed to do the upgrade.

Similarly, do we want to add the a lua migration list that basically would contain every version that this plugin can upgrade from? The idea being if we don't have the correct version then we can short circuit the lua upgrade and not try to go to the python code to upgrade at all. (or the other option is to attempt to upgrade the toolkit using just lua upgrades)

We'd have to make sure that in the case no lua versions exist, but we have a lua toolkit trying to be upgraded to a python plugin, we'd want to check if the schema matches, before we go to python code to migration the rest of the python migrations.

I was gonna bring this up in our meeting monday for more info.

Reply:
same as above. talked about it with the team and filed https://github.com/delphix/virtualization-sdk/issues/115 as well as https://jira.delphix.com/browse/PYT-1338 to address the issues.